src/network/core/packet.h

```diff
@@ -38,49 +40,144 @@ typedef uint8  PacketType; ///< Identifier for the packet
+	 * @param args              The fourth and further parameters to the transfer function, if any.
+	 * @return The return value of the transfer_function.
+	 */
+	template <
+		typename A = size_t, ///< The type for the amount to be passed, so it can be cast to the right type.
+		typename F,          ///< The type of the function.
+		typename D,          ///< The type of the destination.
+		typename ... Args>   ///< The types of the remaining arguments to the function.
+	ssize_t TransferOut(F transfer_function, D destination, Args&& ... args)
+	{
+		return TransferOutWithLimit<A>(transfer_function, std::numeric_limits<size_t>::max(), destination, std::forward<Args>(args)...);
+	}
+
+	/**
+	 * Transfer data from the given function into the packet. It starts writing at the
+	 * position the last transfer stopped.
+	 *
+	 * Examples of functions that can be used to transfer data into a packet are TCP's
+	 * recv and UDP's recvfrom functions. They will directly write their data into the
+	 * packet without an intermediate buffer.
+	 * Examples of functions that can be used to transfer data from a packet are TCP's
+	 * send and UDP's sendto functions. They will directly read the data from the packet's
+	 * buffer without an intermediate buffer.
+	 * These are functions are special in a sense as even though the packet can send or
+	 * receive an amount of data, those functions can say they only processed a smaller
+	 * amount, so special handling is required to keep the position pointers correct.
+	 * Most of these transfer functions are in the form function(source, buffer, amount, ...),
+	 * so the template of this function will assume that as the base parameter order.
+	 *
+	 * This will attempt to write all the remaining bytes into the packet. It updates the
+	 * position based on how many bytes were actually written by the called transfer_function.
+	 * @param transfer_function The function to pass the buffer as second parameter and the
+	 *                          amount to read as third parameter. It returns the amount that
+	 *                          was read or -1 upon errors.
+	 * @param source            The first parameter of the transfer function.
+	 * @param args              The fourth and further parameters to the transfer function, if any.
+	 * @return The return value of the transfer_function.
+	 */
+	template <
+		typename A = size_t, ///< The type for the amount to be passed, so it can be cast to the right type.
+		typename F,          ///< The type of the function.
+		typename S,          ///< The type of the source.
+		typename ... Args>   ///< The types of the remaining arguments to the function.
+	ssize_t TransferIn(F transfer_function, S source, Args&& ... args)
+	{
+		size_t amount = this->RemainingBytesToTransfer();
+		if (amount == 0) return 0;
+
+		/* Making buffer a char means casting a lot in the Recv/Send functions. */
+		char *input_buffer = reinterpret_cast<char*>(this->buffer.data() + this->pos);
```

Comment:
Is there no chance that this->buffer could be smaller than this->pos ? Could cause issues...

Reply:
Upon initialization it is set to 0.
`amount` is defined effectively as `this->buffer.size() - this->pos`, so this function cannot cause it to be bigger.
Then all Recv_u* functions call CanReadFromPacket which checks `this->pos + bytes_to_read <= this->buffer.size()`, and if that fails 0 is returned instead, so those cannot get `this->pos` above `this->buffer.size()`
Recv_string does its own validation against `this->buffer.size()`, so that should not be able to get pos beyond the buffer.
TransferOut has a similar check as TransferIn on the amount.
So I do not know of a code path where it can be the case that pos is bigger than `this->buffer.size()`... except for cosmic rays or buffer-overflow/write-after-free type bugs.

Reply:
How much would it hurt to have an assert to that effect just to make absolutely sure? :)

Reply:
For releases nothing, for the rest probably not much.